src/lib/client.ts

```diff
@@ -0,0 +1,107 @@
+import Bluebird = require("bluebird");
+import _  = require("lodash");
+import request = require("./request");
+
+import Provider from "./providers/provider";
+import PasswordProvider from "./providers/password";
+import OauthProvider from "./providers/oauth";
+
+import Services = require("./services");
+import { ChannelService } from "./services/channel";
+import { ChatService } from "./services/chat";
+import { GameService } from "./services/game";
+
+import { Utils } from "./utils";
+import { UriOptions } from "request";
+import { TeamService } from "./services/team";
+
+export class Client {
+    // Services
+    public channel: ChannelService;
+    public chat: ChatService;
+    public game: GameService;
+    public team: TeamService;
+
+    private provider: Provider = null;
+    private userAgent: string = Utils.getUserAgent();
+    private urls: {
+        api: string;
+        frontend: string;
+    };
+
+    constructor() {
+        this.urls = {
+            api: "https://beam.pro/api/v1",
+            frontend: "https://beam.pro",
+        };
+        _.forIn(Services, (Service, key) => {
+            this[key] = new Service(this);
+        });
+    }
+
+    /**
+     * Sets the API/Frontend URLs for the client.
+     */
+    public setUrl(type: URLType, url: string): void {
+        this.urls[type] = url;
+    }
+
+    /**
+     * Get a URL based on the type passed.
+     */
+    public getUrl(type: URLType): string {
+        return this.urls[type];
+    }
+
+    /**
+     * Creates and returns an authentication provider instance.
+     */
+    public use(provider: "password", options: PasswordOpts): PasswordProvider;
+    public use(provider: "oauth", options: OAuthOpts): OauthProvider;
+    public use(provider: Providers, options: any): any;
+    public use(provider: string, options: any): any {
+        let Provider = require(`./providers/${provider}`).default;
+        this.provider = new Provider(this, options);
+        return this.provider;
+    }
```

Comment:
TypeScript requires you to define all types and the any type to allow the compiler to understand what's going on. Without the "any" variant the definitions won't compile correctly. It's a part bug with TSC it self rather than the spec.

Unless you mean something else?